# Выбор локации для скважины

Нефтяная компания предоставляет пробы нефти из трех регионов, в каждом из которых 10 000 месторождений. С помощью машинного обучения нужно определить, какой регион принесет наибольшую прибыль и проанализировать риски.

## Загрузка и подготовка данных

In [1]:
#импортируем нужные библиотеки
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

In [2]:
#читаем и изучаем данные
try: 
    r_1 = pd.read_csv('/datasets/geo_data_0.csv')
    r_2 = pd.read_csv('/datasets/geo_data_1.csv')
    r_3 = pd.read_csv('/datasets/geo_data_2.csv')
except: 
    r_1 = pd.read_csv('C:\\Users\\Дмитрий\\Desktop\\Проекты Я.Практикум\\oil_production\\geo_data_0.csv')
    r_2 = pd.read_csv('C:\\Users\\Дмитрий\\Desktop\\Проекты Я.Практикум\\oil_production\\geo_data_1.csv')
    r_3 = pd.read_csv('C:\\Users\\Дмитрий\\Desktop\\Проекты Я.Практикум\\oil_production\\geo_data_2.csv')

In [3]:
def info(data):
    data.info()
    display(data.isna().mean())
    display(f'Число явных дубликатов: {data.duplicated().sum()}')
    display(data)

In [4]:
info(r_1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64

'Число явных дубликатов: 0'

id        f0        f1        f2     product
0      txEyH  0.705745 -0.497823  1.221170  105.280062
1      2acmU  1.334711 -0.340164  4.365080   73.037750
2      409Wp  1.022732  0.151990  1.419926   85.265647
3      iJLyR -0.032172  0.139033  2.978566  168.620776
4      Xdl7t  1.988431  0.155413  4.751769  154.036647
...      ...       ...       ...       ...         ...
99995  DLsed  0.971957  0.370953  6.075346  110.744026
99996  QKivN  1.392429 -0.382606  1.273912  122.346843
99997  3rnvd  1.029585  0.018787 -1.348308   64.375443
99998  7kl59  0.998163 -0.528582  1.583869   74.040764
99999  1CWhH  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 5 columns]

In [5]:
info(r_2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64

'Число явных дубликатов: 0'

id         f0         f1        f2     product
0      kBEdx -15.001348  -8.276000 -0.005876    3.179103
1      62mP7  14.272088  -3.475083  0.999183   26.953261
2      vyE1P   6.263187  -5.948386  5.001160  134.766305
3      KcrkZ -13.081196 -11.506057  4.999415  137.945408
4      AHL4O  12.702195  -8.147433  5.004363  134.766305
...      ...        ...        ...       ...         ...
99995  QywKC   9.535637  -6.878139  1.998296   53.906522
99996  ptvty -10.160631 -12.558096  5.005581  137.945408
99997  09gWa  -7.378891  -3.084104  4.998651  137.945408
99998  rqwUm   0.665714  -6.152593  1.000146   30.132364
99999  relB0  -3.426139  -7.794274 -0.003299    3.179103

[100000 rows x 5 columns]

In [6]:
info(r_3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64

'Число явных дубликатов: 0'

id        f0        f1        f2     product
0      fwXo0 -1.146987  0.963328 -0.828965   27.758673
1      WJtFt  0.262778  0.269839 -2.530187   56.069697
2      ovLUW  0.194587  0.289035 -5.586433   62.871910
3      q6cA6  2.236060 -0.553760  0.930038  114.572842
4      WPMUX -0.515993  1.716266  5.899011  149.600746
...      ...       ...       ...       ...         ...
99995  4GxBu -1.777037  1.125220  6.263374  172.327046
99996  YKFjq -1.261523 -0.894828  2.524545  138.748846
99997  tKPY3 -1.199934 -2.957637  5.219411  157.080080
99998  nmxp2 -2.419896  2.417221 -5.548444   51.795253
99999  V9kWn -2.551421 -2.025625  6.090891  102.775767

[100000 rows x 5 columns]

В данных нет пропусков и дубликатов. Столбец `id` не несет информации для прогнозирования и может запутать модель, поэтому удаляем его.

## Подготовка признаков и обучение модели

Создаем функцию для подготовки признаков и моделирования.

In [7]:
def modelling(data, region):
    #удаляем столбец id
    data = data.drop('id', axis=1)

    #отделяем целевой признак и остальные
    features = data.drop('product', axis=1)
    target = data['product']

    #разбиваем данные на обучающую и валидационную выборку
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=12345)

    print(features_train.shape)
    print(features_valid.shape)
    print(target_train.shape)
    print(target_valid.shape)
    
    #обучаем модель линейной регрессии
    model = LinearRegression()
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)

    #оцениваем модель 
    mean_stock = predictions_valid.mean()
    mse = mean_squared_error(target_valid, predictions_valid)
    rmse = mse**0.5
    print(f'Средний запас в регионе {region}: {mean_stock}')
    print(f'RMSE: {rmse}')
    return features_train, features_valid, target_train, target_valid, predictions_valid

In [8]:
features_train_1, features_valid_1, target_train_1, target_valid_1, predictions_valid_1 = modelling(r_1, 1)

(75000, 3)
(25000, 3)
(75000,)
(25000,)
Средний запас в регионе 1: 92.59256778438035
RMSE: 37.5794217150813


In [9]:
features_train_2, features_valid_2, target_train_2, target_valid_2, predictions_valid_2 = modelling(r_2, 2)

(75000, 3)
(25000, 3)
(75000,)
(25000,)
Средний запас в регионе 2: 68.72854689544602
RMSE: 0.8930992867756165


In [10]:
features_train_3, features_valid_3, target_train_3, target_valid_3, predictions_valid_3 = modelling(r_3, 3)

(75000, 3)
(25000, 3)
(75000,)
(25000,)
Средний запас в регионе 3: 94.96504596800489
RMSE: 40.02970873393434


**Выводы**
* Наибольший средний запас скважин в регионе №3: 94.97 тыс.баррелей. Но и ошибка модели также самая высокая: 40.03.
* Чуть ниже средний запас в регионе №1: 92.59 тыс.баррелей. Ошибка равна 37.58.
* Наименьший средний запас в регионе №2: 68.73 тыс.баррелей. Однако модель очень точна, ошибка равна всего 0.89.

## Подготовка к расчёту прибыли

In [11]:
#рассчет точки безубыточности скважины
N=200  #число скважин
BUDGET = 10000000 #бюджет на разработку 200 скважин (тыс рублей)
COST = 450 #цена тысячи баррелей (тыс рублей)
PRODUCTION = BUDGET/(N*COST)
print(f'Безубыточность скважины возможна при объеме не менее {PRODUCTION} тыс.баррелей')


Безубыточность скважины возможна при объеме не менее 111.11111111111111 тыс.баррелей


Исходя из среднего объема скважин, все регионы будут убыточными. Однако отбираться будут только 200 лучших скважин, и для нас важно среднее по ним, а не всему региону.

## Расчёт прибыли и рисков 

In [12]:
#создаем функцию для подсчета прибыли
def profit(target, predictions, count):
    pred_sorted = predictions.sort_values(ascending=False)
    selected = target[pred_sorted.index][:count]
    return (selected.sum()*COST - BUDGET)/1000

In [13]:
#создаем функцию для бутстрепа
state = np.random.RandomState(12345)
    
def result(target, predictions):
    values=[] 
    for i in range(1000):
        target_subsample = target.sample(n=500, replace=True, random_state=state)
        probs_subsample = predictions[target_subsample.index]
        result = profit(target_subsample, probs_subsample, 200)
        values.append(result)
    values = pd.Series(values)
    percentage_of_negative = (values < 0).mean()*100
    mean = values.mean()
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)
    print(f'Средняя прогнозная прибыль равняется {mean} млн рублей')
    print(f'С вероятностью 95% прибыль составит от {lower} до {upper} млн рублей')
    print(f'Риск убытков составляет {percentage_of_negative}%')

In [14]:
#применяем функции к данным
target_valid_1 = target_valid_1.reset_index(drop=True)
predictions_valid_1 = pd.Series(predictions_valid_1)
result(target_valid_1, predictions_valid_1)

Средняя прогнозная прибыль равняется 425.9385269105924 млн рублей
С вероятностью 95% прибыль составит от -102.09009483793722 до 947.9763533583689 млн рублей
Риск убытков составляет 6.0%


In [15]:
target_valid_2 = target_valid_2.reset_index(drop=True)
predictions_valid_2 = pd.Series(predictions_valid_2)
result(target_valid_2, predictions_valid_2)

Средняя прогнозная прибыль равняется 518.2594936973248 млн рублей
С вероятностью 95% прибыль составит от 128.1232314330842 до 953.6129820669086 млн рублей
Риск убытков составляет 0.3%


In [16]:
target_valid_3 = target_valid_3.reset_index(drop=True)
predictions_valid_3 = pd.Series(predictions_valid_3)
result(target_valid_3, predictions_valid_3)

Средняя прогнозная прибыль равняется 420.1940053440501 млн рублей
С вероятностью 95% прибыль составит от -115.85260916001162 до 989.629939844574 млн рублей
Риск убытков составляет 6.2%


## Вывод

Наиболее перспективный - регион №2, т.к. он имеет наибольшую среднюю прогнозную прибыль (518 млн рублей) и практически нулевой риск убытков (0.3%). В остальных регионах риск убытков превышает 2.5%. 